src/state/opamSysInteract.ml

```diff
@@ -0,0 +1,420 @@
+  | "suse" | "opensuse" ->
+    ["zypper"::y_answer ["--non-interactive"] ("install"::packages)]
+  | family ->
+    OpamConsole.error_and_exit `Not_found
+      "opam doesn't handle system %s for system packages install commands.\n \
+       Re-run your command with `--i-am-not-root` option to bypass this step"
+      family
+
+let update_command () =
+  match spv OpamSysPoll.os_family with
+  | "debian" ->
+    Some ["apt-get";"update"]
+  | "homebrew" ->
+    Some ["brew"; "update"]
+  | "rhel" | "centos" | "fedora" | "mageia" | "oraclelinux" | "ol" ->
+    Some ["yum"; "makecache"]
+  | "archlinux" | "arch" ->
+    Some ["pacman"; "-Sy"]
+  | "gentoo" ->
+    Some ["emerge"; "--sync"]
+  | "alpine" ->
+    Some ["apk"; "update"]
+  | "suse" | "opensuse" ->
+    Some ["zypper"; "--non-interactive"; "update"]
+  | _ -> None
+
+let sudo_run_command cmd =
+  (* Allow it as option ? *)
+  let su = OpamSystem.resolve_command "sudo" = None in
+  let get_cmd = function
+    | c::a -> c, a
+    | _  -> assert false
+  in
+  let check_sudo sudo cmd =
+    OpamConsole.warning
+      "The following command needs to be run through %S:\n    %s\n%!"
+      sudo  (String.concat " " cmd);
+    assert OpamStateConfig.(not (!r.depext_no_sudo || !r.depext_not_root));
+    if not (OpamConsole.confirm "Allow & continue?") then
+      OpamStd.Sys.exit_because `Aborted
+  in
+  let cmd, args =
+    match spv OpamSysPoll.os, spv OpamSysPoll.os_distribution with
+    | "openbsd", _ ->
+      if Unix.getuid () <> 0 then (
+        check_sudo "doas" cmd;
+        "doas", cmd
+      ) else get_cmd cmd
+    | ("linux" | "unix" | "freebsd" | "netbsd" | "dragonfly"), _
+    | "macos", "macports" ->
+      if Unix.getuid () <> 0 then (
+        check_sudo (if su then "su" else "sudo") cmd;
+        if su then
+          "su", ["root"; "-c"; Printf.sprintf "%S" (String.concat " " cmd)]
+        else
+          "sudo", cmd
+      ) else get_cmd cmd
+    | _ -> get_cmd cmd
+  in
+  let code = run_command_exit_code cmd args in
+  if code <> 0 then Some (cmd,code)
+  else None
+
+let update () =
+  match update_command () with
+  | Some cmd ->
+    (match sudo_run_command cmd with
+     | Some (cmd, code) ->
+       OpamConsole.error_and_exit `False
+         "System update failed with exit code %d: %s" code cmd
+     | None ->
+       OpamConsole.note "System updated succesfully")
+  | None ->
+    OpamConsole.warning "Unknown system %s, skipping system update"
+      (spv OpamSysPoll.os_family)
+
+let install packages =
+  if OpamSysPkg.Set.is_empty packages then ()
+  else
+  let cmds =
+    install_packages_commands packages
+  in
+  let ok =
+    List.fold_left (fun ok cmd ->
+        match ok with
+        | None -> sudo_run_command cmd
+        | Some _ as s -> s)
+      None cmds
+  in
+  match ok with
+  | Some (cmd, code) ->
+    OpamConsole.error_and_exit `False
+      "System packages install failed with exit code %d at command:\n  %s"
+      code cmd
```

Comment:
This can be improved later on, but I am not sure this should be a hard failure...
we could imagine e.g. to pause and let the user try to fix it, then retry.

Reply:
agree!

Reply:
updated